### ISSUE TITLE
Should dotnet `latest` show if it is prior to the current version?

### ISSUE THREAD

**Issue by eamodio** [vscode-contrib/vscode-versionlens#65]
Feels a bit misleading -- even though it is technically correct, since it is the latest non-prerelease.

![image](https://cloud.githubusercontent.com/assets/641685/26028662/70989322-37f3-11e7-9351-4b723a101e05.png)

**Comment by pflannery** [comment]
I think it should still show the latest tag but have a down arrow instead

Like this
![image](https://cloud.githubusercontent.com/assets/1727302/26029113/115b0ee4-3825-11e7-91be-c9e1a43b8a00.png)

**Comment by eamodio** [comment]
That would certainly be better -- but still a bit hard to parse as the arrow is quite small and the difference is not large. Maybe if the arrow was at the end rather than the start there would be a bigger visual difference? 

To be honest though I think this is part of a bigger issue, at least imo. There just feels like a lot going on in the code lens. Maybe if code lens weren't rendered at all when the match and the latest are the same? Maybe the `Matches {version}` code lens only said `Matches` since the version is kind of redundant or maybe `Up to date`? I dunno.

I started thinking about this and here is a strawman proposal to hopefully simplify the ux:

- if current version not prerelease version, and:
  - is the latest version, then:
    - show no code lens or maybe `Up to date` (my pref would be no code lens so easier to scan for actionable things)
  - is NOT the latest version, and:
    - the latest version is in range, then:
       - show an `Update to {version}` code lens to the latest release in range (maybe with a sync icon)
       - if there is also a later prerelease version, then:
         - also show an `Upgrade to {version}` code lens to the latest prerelease (maybe with an up arrow)
    - the latest is in not in range, then:
       - show an `Upgrade to {version}` code lens to the latest release (maybe with an up arrow)
       - if there is also a later prerelease version, then:
         - also show an `Upgrade to {version}` code lens to the latest prerelease (maybe with an up arrow)

- if current version a prerelease, and:
  - Always show a `Downgrade to {version}` code lens to the latest non-prerelease (maybe with a down arrow)
  -  is the latest version, then:
    - show no code lens or maybe `Up to date` (my pref would be no code lens so easier to scan for actionable things)
  - is NOT the latest version, and:
    - the latest version is in range, then:
       - show an `Update to {version}` code lens to the latest release in range (maybe with a sync icon)
       - if there is also a later prerelease version, then:
         - also show an `Upgrade to {version}` code lens to the latest prerelease (maybe with an up arrow)
    - the latest is in not in range, then:
       - show an `Upgrade to {version}` code lens to the latest release (maybe with an up arrow)

I don't think that covers 100% of the cases, but hopefully you get idea. What do you think?

**Comment by pflannery** [comment]
I've also been thinking about changing "Matches" but I want to change it to "Installs <x.x.x>" to make it clearer what you would be installing with the version range chosen.

I will look at dropping the latest code lens entry when the latest is already chosen. Then just present it as "Installs latest". 

Problem with things like "Downgrade to" or "Upgrade to" etc is we don't really know whats installed and it's all in theoretical space.

I definitely want to keep code lenses rendered as the lines become disjointed without them.

I would like to get the outdated decorations working for dot net in the future which will give a better visual indication of whats out of date.

**Comment by eamodio** [comment]
WRT to "Installs {version}" you aren't really installing anything at that point though right? You are just updating a version number that will at some point later resolve to a particular package version (or a minimum package version). Because of that, I've just been thinking about what is being done to that version number -- `update` - increase to later version within a predefined acceptable range, `upgrade` - increase to a later version outside that predefined range, `downgrade` - decrease to a prior version. None of those will reflect the state of what is actually installed, just what the state will be after an external "install, update, restore" is run utilizing the versions specified.

I also like the separation (or at least my perceived separation) of the code lens vs the decorations. The code lens in my mind are all about the "desired" version (or range) -- like in the context of a clean restore or CI. While the decorations reflect the current state of my local environment at that point in time, against the "desired" version (aside -- imo it would be nice to have a setting to turn off the decorations).

Anyway just my 2c ;)

**Comment by pflannery** [comment]
@eamodio  

[pre-releases/0.18.0-beta.1](https://github.com/vscode-contrib/vscode-versionlens/releases/tag/0.18.0-beta.1) contains some clearer matching reasons.

It isn't published to vscode yet but you can try it by downloading the vsix file and installing manually

**Comment by eamodio** [comment]
@pflannery Looks awesome! Huge improvement! I only have opinionated issues so certainly ignore them as you see fit ;)

I just find `Matches the latest` -- quite a lot to read / distracting since there is nothing to act upon. Also wonder what you think of using `pinned` rather than `fixed` feel clearer to me.

Maybe something like these?

![image](https://cloud.githubusercontent.com/assets/641685/26114834/5ab38202-3a2c-11e7-9fac-87b5f0698a09.png)

![image](https://cloud.githubusercontent.com/assets/641685/26114977/c2bf8de6-3a2c-11e7-8e99-c2d90b448204.png)

Again just my 2c ;)
